include/range/v3/experimental/utility/generator.hpp

```diff
@@ -0,0 +1,353 @@
+// Range v3 library
+//
+//  Copyright Casey Carter 2017
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+#ifndef RANGES_V3_EXPERIMENTAL_UTILITY_GENERATOR_HPP
+#define RANGES_V3_EXPERIMENTAL_UTILITY_GENERATOR_HPP
+
+#include <range/v3/detail/config.hpp>
+#if RANGES_CXX_COROUTINES >= RANGES_CXX_COROUTINES_TS1
+#include <cstddef>
+#include <exception>
+#include <utility>
+#include <experimental/coroutine>
+#include <meta/meta.hpp>
+#include <range/v3/range_traits.hpp>
+#include <range/v3/view_facade.hpp>
+#include <range/v3/detail/optional.hpp>
+#include <range/v3/utility/box.hpp>
+#include <range/v3/utility/concepts.hpp>
+#include <range/v3/utility/swap.hpp>
+#include <range/v3/view/all.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        /// \addtogroup group-utility
+        /// @{
+        namespace experimental
+        {
+            // The type of size() for a sized_generator
+            using generator_size_t = std::size_t;
+
+            // Type upon which to co_await to set the size of a sized_generator
+            enum struct generator_size : generator_size_t { invalid = ~generator_size_t(0) };
+
+            template<typename Promise = void>
+            struct coroutine_owner;
+        } // namespace experimental
+
+        /// \cond
+        namespace detail
+        {
+            inline void resume(std::experimental::coroutine_handle<> coro)
+            {
+                // Pre: coro refers to a suspended coroutine.
+                RANGES_EXPECT(coro);
+                RANGES_EXPECT(!coro.done());
+                coro.resume();
+            }
+
+            namespace coroutine_owner_
+            {
+                struct adl_hook {};
+
+                template<typename Promise>
+                void swap(experimental::coroutine_owner<Promise> &x,
+                          experimental::coroutine_owner<Promise> &y) noexcept
+                {
+                    ranges::swap(x.base(), y.base());
+                }
+            } // namespace coroutine_owner_
+        } // namespace detail
+        /// \endcond
+
+        namespace experimental
+        {
+            // An owning coroutine_handle
+            template<typename Promise>
+            struct coroutine_owner
+            : std::experimental::coroutine_handle<Promise>
```

Comment:
Dammit - VS Code broke all of these partial indent lines when I moved this stuff in an indent for the `experimental` namespace, and I missed it.